main.go

```diff
@@ -474,8 +520,111 @@ func makeRequestWithTracePriority(t transport.Transport, request *transport.Requ
+	}
+	return false
+}
+
+// receiveStreamMessage receives and prints the stream message response
+// returns true if stream ends with EOF
+func receiveStreamMessage(ctx context.Context, stream *yarpctransport.ClientStream, serializer encoding.Serializer, out output) bool {
+	msg, err := stream.ReceiveMessage(ctx)
+	if err == io.EOF {
+		return true
+	}
+	if err != nil {
+		out.Fatalf("Failed while receiving stream response: %v\n", err)
+	}
+
+	bytes, err := ioutil.ReadAll(msg.Body)
+	if err != nil {
+		out.Fatalf("Failed while reading stream response: %v\n", err)
+	}
+	res, err := serializer.Response(&transport.Response{Body: bytes})
+	if err != nil {
+		out.Fatalf("Failed while serializing stream response: %v\n", err)
+	}
+
+	bs, err := json.MarshalIndent(res, "", "  ")
+	if err != nil {
+		out.Fatalf("Failed to convert map to JSON: %v\nMap: %+v\n", err, res)
+	}
+	out.Printf("%s\n\n", bs)
+	return false
+}
+
+func closeStream(ctx context.Context, stream *yarpctransport.ClientStream, out output) {
+	if err := stream.Close(ctx); err != nil {
+		out.Fatalf("Failed to close send stream: %v\n", err)
+	}
+}
+
+func makeInitialStreamRequest(out output, t transport.Transport, serializer encoding.Serializer, streamMsgReader encoding.StreamRequestReader, req *transport.Request) {
+	streamTransport, ok := t.(transport.StreamTransport)
+	if !ok {
+		out.Fatalf("Transport does not support stream calls: %q", t.Protocol())
+	}
+	streamSerializer, ok := serializer.(encoding.StreamSerializer)
+	if !ok {
+		out.Fatalf("Serializer does not support stream calls: %q", serializer.Encoding())
+	}
+
+	ctx, cancel := tchannel.NewContext(req.Timeout)
+	defer cancel()
+	ctx = makeContextWithTrace(ctx, t, req, 0)
+	stream, err := streamTransport.CallStream(ctx, req)
+	if err != nil {
+		out.Fatalf("Failed while making stream call: %v\n", err)
+	}
+
+	if streamSerializer.IsClientStreaming() && streamSerializer.IsServerStreaming() {
+		// bi-directional stream
+		for {
+			if eof := sendStreamMessage(ctx, stream, streamMsgReader, out); eof {
+				closeStream(ctx, stream, out)
+				break
+			}
+			if eof := receiveStreamMessage(ctx, stream, serializer, out); eof {
+				out.Fatalf("Received EOF while receiving bi-directional stream response: %v\n", io.EOF)
+			}
+		}
+	} else {
+		serverStreaming := streamSerializer.IsServerStreaming()
+		clientStreaming := streamSerializer.IsClientStreaming()
+		for {
+			eof := sendStreamMessage(ctx, stream, streamMsgReader, out)
+			// send message only once if the method is server streaming
+			// since server streaming method expects only one request
+			if eof || serverStreaming {
+				break
+			}
+		}
+		closeStream(ctx, stream, out)
+		for {
+			eof := receiveStreamMessage(ctx, stream, serializer, out)
+			// receive message only once if the method is client streaming
+			// since client streaming method expects only one response
+			if eof || clientStreaming {
+				break
+			}
+		}
+	}
```

Comment:
I have few questions about this block of codes:

- In the bi-directional stream mode, we will be in a mode where we send_a_message/get_a_message/send_a_message/get_a_message etc. Is it what we want? This is a very predictable way of streaming but this is not always the case. Should we have concurrent read and write here?
- In the design part of this feature, we had an option called `stream-interval` for sending messages from the client. Where are we going to use this one?

Reply:
* Assumption to have predictable stream in initial request was to enable simple interactive bidirectional streaming i.e.
```
{req-1} // user-input
{res-1} // server-response
{req-2} // user-input
{res-2} // server-response
...
```
Your point is valid though, we might have bidi-streams sending non-uniform responses.

* `stream-interval` will be used for non-interactive mode of request input, I'd prefer adding additional options in a separate PR once these changes are good to merge